feat: Add Insta Snapshot Testing

kutal10 opened the pull request:
## Summary

Introduces snapshot testing via the `insta` crate and refactors tests accordingly. It keeps the Rust toolchain version at 1.80 to meet the downstream requirement of
[Apache OpenDAL](https://github.com/apache/opendal)

This PR really focuses on improving clarity & maintainability

Resolves https://github.com/fast/fastrace/issues/66

## Changes

### 1. Added `insta` for Snapshot Testing
- Introduces [`insta`](https://crates.io/crates/insta) as a testing dependency.
- Enables diff-based snapshot review on assertion failure.
- Snapshots are managed via the CLI (`cargo insta review`).

### 2. Refactored Tests to Use Snapshots
- Converted relevant test assertions to snapshot-based assertions using `insta`.

---

## Testing

```powershell
$ cargo test --features snapshots  

test result: ok. 55 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 4.01s
```

## Documentation 

The `CONTRIBUTING.md` has been updated to take note of how to run all relevant tests with the snapshots feature flag.

andylokandy commented:
Hey @kutal10! Thanks so much for this contribution. The snapshot testing implementation looks really clean and well thought out.

One thing I'm a bit concerned about is how this might affect our MSRV. We're currently locked at 1.80 because of our major dependant [apache/opendal](https://github.com/apache/opendal). Have you checked if insta is compatible with 1.80?

Also, I was wondering if you'd considered using inline snapshots instead? They have the advantage of keeping the test content right next to the test logic, which I find makes the tests easier to understand and maintain. You wouldn't need to jump between files to see what's being tested.

Otherwise, great work on this!  Let me know your thoughts on these suggestions.

kutal10 commented:
> One thing I'm a bit concerned about is how this might affect our MSRV. We're currently locked at 1.80 because of our major dependant [apache/opendal](https://github.com/apache/opendal). Have you checked if insta is compatible with 1.80?

Yep Insta works fine with 1.80, turns out the bump was just to help resolve something locally in my toolchain. Updated the PR to make mention it maintains the toolchain at 1.80. 

> Also, I was wondering if you'd considered using inline snapshots instead? They have the advantage of keeping the test content right next to the test logic, which I find makes the tests easier to understand and maintain. You wouldn't need to jump between files to see what's being tested.

Makes sense, it's a fairly self contained file. I've updated it to be inline now! 

PR's been updated - CI passes!

andylokandy commented:
What is the reason why snapshot tests need the 'snapshot' feature flag? AFAIN, insta would work as normal test unless we need to update the snapshot.

kutal10 commented:
@andylokandy 

Great question so you're right - insta works as normal if you just run `cargo test` if I removed the feature flag attributes. However, I  believes it's good practice to keep:

- If snapshot tests are failing because snapshots weren’t updated or `cargo insta review` wasn’t run, it could possibly break builds or introduce confusion. 
- It also signals "These tests are tied to output graph checks and may require developer review or snapshot updates."

So it's more of a clean separation of concerns to segregate logic tests vs specialised output tests. I have made a choice to make snapshot testing opt-in and explicit for just developer intent purposes. 

It wouldn't impact downstream consumers as `Insta` is a dev dependency of course

andylokandy commented:
> If snapshot tests are failing because snapshots weren’t updated or cargo insta review wasn’t run, it could possibly break builds or introduce confusion.

If someone changes code that affects snapshots, they'll need the insta tools installed or CI will break. Maybe we should document this requirement clearly?

kutal10 commented:
Just so I understand, you'd rather I remove the feature flag, have it run every time, and instead update docs to make it clear that you need `cargo-insta` installed locally in the event any relevant snapshot tests change?

andylokandy commented:
lgtm

kutal10 commented:
@andylokandy Aforementioned changes pushed!

andylokandy commented on CONTRIBUTING.md:
```diff
@@ -9,14 +9,22 @@ This is a pure rust project, so only `cargo` is needed.
 - `cargo build` to compile the current package.
 - `cargo clippy` to catch common mistakes and improve code.
-- `cargo test` to run unit tests.
+- `cargo test` to run unit tests*.
```
is this a typo?

kutal10 replied:
Was not a typo - was using the asterisk as a footnote. Updated to be more explicit https://github.com/fast/fastrace/pull/95/commits/476316f44d43ba9fd29cbb427ea30f403220d3a9

andylokandy commented on CONTRIBUTING.md:
```diff
@@ -9,14 +9,22 @@ This is a pure rust project, so only `cargo` is needed.
 - Check/Build/Test/Clippy all code: `cargo <cmd> --all-targets --workspace`
 - Test specific function: `cargo test multiple_local_parent`
 
+### *Testing 
```
ditto

kutal10 replied:
Was not a typo - was using the asterisk as a footnote. Updated to be more explicit https://github.com/fast/fastrace/pull/95/commits/476316f44d43ba9fd29cbb427ea30f403220d3a9

andylokandy commented:
LGTM. Thank you very much!
